bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,463 @@
+            }
+        }
+    }
+
+    private class IncomingMessageHandler extends AbstractIqRequestHandler {
+        private final EventListener listener;
+        private final ReportParser parser;
+        private final ConcurrentHashMap<String, CommandResponseHolder> pendingCommands = new ConcurrentHashMap<>();
+        private boolean disposed;
+
+        IncomingMessageHandler(EventListener listener) {
+            super("query", "com:ctl", Type.set, Mode.async);
+            this.listener = listener;
+            this.parser = new XmlReportParser(EcovacsXmppDevice.this, listener, gson);
+        }
+
+        void registerPendingCommand(String id, CommandResponseHolder responseHolder) {
+            pendingCommands.put(id, responseHolder);
+        }
+
+        void unregisterPendingCommand(String id) {
+            pendingCommands.remove(id);
+        }
+
+        void dispose() {
+            disposed = true;
+        }
+
+        @Override
+        public @Nullable IQ handleIQRequest(@Nullable IQ iqRequest) {
+            if (disposed) {
+                return null;
+            }
+
+            if (iqRequest instanceof DeviceCommandIQ) {
+                DeviceCommandIQ iq = (DeviceCommandIQ) iqRequest;
+
+                try {
+                    if (!iq.id.isEmpty()) {
+                        CommandResponseHolder responseHolder = pendingCommands.remove(iq.id);
+                        if (responseHolder != null) {
+                            synchronized (responseHolder) {
+                                responseHolder.response = iq.payload;
+                                responseHolder.notifyAll();
+                            }
+                        }
+                    } else {
+                        Optional<String> eventNameOpt = XPathUtils.getFirstXPathMatchOpt(iq.payload, "//ctl/@td")
+                                .map(n -> n.getNodeValue());
+                        if (eventNameOpt.isPresent()) {
+                            logger.trace("{}: Received event message XML {}", getSerialNumber(), iq.payload);
+                            parser.handleMessage(eventNameOpt.get(), iq.payload);
+                        } else {
+                            logger.debug("{}: Got unexpected XML payload {}", getSerialNumber(), iq.payload);
+                        }
+                    }
+                } catch (Exception e) {
+                    listener.onEventStreamFailure(EcovacsXmppDevice.this, e);
+                }
+            } else if (iqRequest instanceof ErrorIQ) {
+                StanzaError error = ((ErrorIQ) iqRequest).getError();
+                logger.trace("{}: Got error response {}", getSerialNumber(), error);
+                listener.onEventStreamFailure(EcovacsXmppDevice.this,
+                        new XMPPException.XMPPErrorException(iqRequest, error));
+            }
+            return null;
```

Comment:
Maybe I am missing something, or it is just not expected by the communication partner, but *not* sending an IQ response would be an violation of the XMPP.

Reply:
But this request handler actually is the response handler ;-) More on that below...